library/src/test/java/com/getstream/sdk/chat/rest/ClientTest.java

```diff
@@ -0,0 +1,278 @@
+
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+/*
+ * Created by Anton Bevza on 2019-10-23.
+ */
+public class ClientTest {
+
+    private static String TEST_API_KEY = "testApiKey";
+    private static String TEST_CLIENT_ID = "testClientId";
+
+    private MockWebServer mockWebServer;
+
+    @Spy
+    private TestTokenProvider testTokenProvider = new TestTokenProvider();
+    private Client client;
+
+    @BeforeEach
+    void initTest() throws IOException {
+        MockitoAnnotations.initMocks(this);
+        WebSocketServiceProvider webSocketServiceProvider = Mockito.mock(WebSocketServiceProvider.class);
+        Mockito.doReturn(new TestWebSocketService(null, "", null))
+                .when(webSocketServiceProvider).provideWebSocketService(
+                Mockito.any(), Mockito.any(), Mockito.any()
+        );
+        mockWebServer = new MockWebServer();
+        mockWebServer.start();
+        TestApiClientOptions testApiClientOptions =
+                new TestApiClientOptions(mockWebServer.url("/").toString());
+        client = new Client(TEST_API_KEY,
+                new ApiServiceProvider(testApiClientOptions),
+                webSocketServiceProvider,
+                new UploadStorageProvider(testApiClientOptions),
+                null);
+    }
+
+    @Test
+    void updateChannelSuccessTest() throws IOException, InterruptedException {
+        simulateConnection();
+        mockWebServer.enqueue(new MockResponse().setResponseCode(200).setBody(
+                new MockResponseFileReader("channel_success.json").getContent()
+        ));
+        final Object syncObject = new Object();
+        ChannelCallback callback = new ChannelCallback() {
+            @Override
+            public void onSuccess(ChannelResponse response) {
+                assertNotNull(response.getChannel());
+                synchronized (syncObject) {
+                    syncObject.notify();
+                }
+            }
+
+            @Override
+            public void onError(String errMsg, int errCode) {
+
+            }
+        };
+        HashMap<String, Object> extraFields = new HashMap<>();
+        extraFields.put("name", "testName");
+        extraFields.put("title", "testTitle");
+        Channel channel = new Channel(client, "testType", "testId", extraFields);
+        Message message = new Message();
+        message.setText("test message");
+        client.updateChannel(channel, message, callback);
+        RecordedRequest recordedRequest = mockWebServer.takeRequest();
+        assertEquals("/channels/testType/testId?api_key="
+                + TEST_API_KEY + "&client_id=" + TEST_CLIENT_ID, recordedRequest.getPath());
+        String expectedBody = "{\"data\":{\"name\":\"testName\",\"title\":\"testTitle\"}," +
```

Comment:
>Again there is no need to how Gson works. Although it's good idea to test data as it can't be tested at compile time by retrofit.

Yeah, you are right. I added the checking of Gson because we use custom TypeAdapter and Tommaso wanted to check is requests build in the right way. But anyway it will be better to test TypeAdapters separately.

Reply:
>And all tests will look about like this:

>fun test(){
  val api = mock(APIService.class)
  val client = Client(api)
  val callback = Callback()  
  when(api.someMethod( params )).return( response )
  client.someMethod( params )
 verify(api).someMethod( params )
 verify(callback).success()
}

It's looks very good. Unfortunately, Client needs totally refactor to write tests like that. I will try to do something to have test the most close to it.

Reply:
So my example maybe too perfect for now. But I think if we at least pass `APIService` interface to the `Client` constructor it might improve tests a lot.

Callbacks will be still there and other client dependencies initialisation.